### ISSUE TITLE
Alternating error on database rebuilds: (error "Selecting deleted buffer")

### ISSUE THREAD

**Issue by dansanduleac** [org-roam/org-roam#1289]
### Description

I'm hitting the exact same problem as #859 on Windows 10 with Spacemacs. If exact version is important, I'm on Spacemacs v.0.300.0 (develop 7a9a16ec9).

I'm using a sqlite3 from the android sdk.

https://github.com/org-roam/org-roam/issues/859#issuecomment-647486944:
> Could you try `M-x org-roam-db-build-cache` twice in a row? I expect that the first call returns quickly with the “selected deleted buffer” message, but the immediate second one works. This is after you delete the db file.

This behaves exactly as you predicted.

#### Steps to Reproduce

1. Load Emacs
3. run the `org-roam-db-build-cache` command - fails as below
4. run the `org-roam-db-build-cache` command again - this time it succeeds

#### Backtrace

```
Debugger entered--Lisp error: (error "Selecting deleted buffer")
  #f(compiled-function (conn) #<bytecode 0x424f7a1>)(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908>)
  apply(#f(compiled-function (conn) #<bytecode 0x424f7a1>) #<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908> nil)
  emacsql-waiting-p(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908>)
  #f(compiled-function (connection &optional timeout) "Block until CONNECTION is waiting for further input." #<bytecode 0x4249f6d>)(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908>)
  apply(#f(compiled-function (connection &optional timeout) "Block until CONNECTION is waiting for further input." #<bytecode 0x4249f6d>) #<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908> nil)
  emacsql-wait(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908>)
  #f(compiled-function (connection sql &rest args) "Send SQL s-expression to CONNECTION and return the results." #<bytecode 0x424b7a1>)(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908> [:select * :from files])
  apply(#f(compiled-function (connection sql &rest args) "Send SQL s-expression to CONNECTION and return the results." #<bytecode 0x424b7a1>) #<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908> [:select * :from files])
  emacsql(#<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908> [:select * :from files])
  apply(emacsql #<emacsql-sqlite3-connection emacsql-sqlite3-connection-542c908> [:select * :from files] nil)
  org-roam-db-query([:select * :from files])
  org-roam-db--get-current-files()
  org-roam-db-build-cache(nil)
  funcall-interactively(org-roam-db-build-cache nil)
  call-interactively(org-roam-db-build-cache record nil)
  command-execute(org-roam-db-build-cache record)
  counsel-M-x-action("org-roam-db-build-cache")
  ivy-call()
  ivy-read("M-x " ("org-roam-build-cache" "eval-expression" "execute-extended-command" "spacemacs/display-and-copy-version" "spacemacs/echo" "emacs-version" "spacemacs/toggle-debug-on-error" "restart-emacs" "org-roam-mode" "org-roam-db-clear" "org-roam-db--clear" "package-reinstall" "desktop-clear" "toggle-debug-on-error" "org-roam-version" "sql-sqlite" "open-line" "org-ctrl-c-ret" "org-tags-sparse-tree" "org-sparse-tree" "evil-indent" "fill-region" "fill-paragraph" "align" "align-current" "eval-region" "eval-defun" "evil-redo" "help-go-back" "describe-symbol" "completion-help-at-point" "debugger-continue" "package-install" "org-roam" "forward-sexp" "make-directory" "complete-symbol" "define-word-at-point" "ivy-spacemacs-help-layers" "configuration-layer/create-layer" "dotspacemacs/sync-configuration-layers" "debug" "widen" "version" "diff-mode" "load-file" "counsel-cl" "other-frame" "eval-last-sexp" "emacs-lisp-mode" ...) :predicate #f(compiled-function (x) #<bytecode 0x17ca0ad>) :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  counsel-M-x()
  funcall-interactively(counsel-M-x)
  call-interactively(counsel-M-x nil nil)
  command-execute(counsel-M-x)
```

### Environment
- Emacs: GNU Emacs 27.0.91 (build 1, x86_64-w64-mingw32)
 of 2020-04-20
- Framework: Spacemacs
- Org: Org mode version 9.4 (9.4-31-g49f4c3-elpaplus @ c:/Users/Dan/AppData/Roaming/.emacs.d/elpa/27.0/develop/org-plus-contrib-20201109/)
- Org-roam: 1.2.3

**Comment by jethrokuan** [comment]
Evaluate this:

```
(defun emacsql-sqlite3--proc-sentinel (proc _change)
  "Called each time when PROC status changed."
  (unless (process-live-p proc)
    (let ((buf (process-buffer proc)))
      (when (buffer-live-p buf)
        (switch-to-buffer buf)))))
```

And paste what's in the resultant buffer upon seeing the error.

**Comment by notuntoward** [comment]
I see a problem similar to the one reported by @dansanduleac:

1. Load Emacs
2. run the org-roam-db-build-cache command - it fails
3. run the org-roam-db-build-cache command again - it succeeds

The sequence above results in a  "Selecting Deleted Buffer" messsage in the minibuffer after running step 2.  And after step 3, I get the normal message like:

> (org-roam) total: Δ33, files-modified: Δ0, ids: Δ0, links: Δ0, tags: Δ0, titles: Δ0, refs: Δ0, deleted: Δ0

I also see the "Selecting deleted buffer" messages at random times while doing org-roam operations.

If I add the sentinel code suggested above to my .emacs, and try again, I get a timeout warning.  Here's the relevant part of what's in *Messages* after steps 2. and 3. above

> emacsql-sqlite3--proc-sentinel
> Query timeout error: "Query timed out", 30
> (org-roam) total: Δ33, files-modified: Δ0, ids: Δ0, links: Δ0, tags: Δ0, titles: Δ0, refs: Δ0, deleted: Δ0

**Comment by dansanduleac** [comment]
Sorry for the delay, here's my attempt as well - I also get a timeout on that function.

```
Error running timer ‘org-roam-db-update-cache-on-timer’: (emacsql-timeout "Query timed out" 30)
Quit
(org-roam) Processed 0/1 modified files...
(org-roam) total: Δ5, files-modified: Δ1, ids: Δ0, links: Δ0, tags: Δ0, titles: Δ1, refs: Δ0, deleted: Δ0
```

and the *emacsql sqlite* buffer that pops up is empty.

**Comment by notuntoward** [comment]
@jethrokuan and @zaeph , when I save an org-roam note, I get a slightly different error message:

> Error running timer ‘org-roam-db-update-cache-on-timer’: (error "Selecting deleted buffer")

I get this message the 1st time I try to save; the 2nd time, I don't get it

**Comment by nobiot** [comment]
Please allow me to offer a pragmatic approach from my experience (you could call it a workaround) to this issue reported. 

I use Org-roam on Windows 10 (vanilla Emacs).

You have two "cases" where you get the `"Selecting deleted buffer"` error.

1. When you run `org-roam-db-build-cache` (either manually, or automatically when you start `org-roam-mode`). The odd-number of times (first, third, 5th, and so on) are always success. The even-number of times are always errors (second, 4th, 6th, and so on). 

   For this, I would say... Live with it. As the first is success when `org-roam-mode` starts, the error is usually not a big issue -- if you need to build the DB cache manually, and get an error, simply run it again immediately after you have seen the error

2. When you are writing a note, and save the buffer, you get the error at a timing that appears to be random (it's actually after 2 seconds of idle time by default). This is much more annoying, and gets in your way of writing. 

   For this, I suggest Windows users to try to set `org-roam-db-update-method` to `'immediate`. This way, Org-roam tries to update the DB cache immediately after you save the buffer -- instead of queueing the update to asynchronously run the DB update. For some reason I cannot explain, the immediate updates give me no `Selecting deleted buffer` error (!) so far. For Windows machine I use, the DB update process is almost instantaneous, so it does not slow me down in writing note (and I save a lot, almost for every sentence I write; I also have an automatic save turned on. I experience no slow down).

I use the following versions of Emacs and packages from MELPA (they should latest, or very close to it).
- emacsql-sqlite3 20200914.508. 
- org-roam 20201124.1309
- Emacs 27.1
- Org Mode 9.4

---

Just a little of background why I suggests the approach.

The reported issue of getting`(error "Selecting deleted buffer")` when you run `org-roam-db-build-cache` has been a common issue since almost the beginning of Org-roam. 

As far as I remember when we needed to compile `emacsql-sqlite` from the source -- later compilation became unnecessary because of the switch to `emacsql-sqlite3`.

It's likely to be a problem somewhere upstream (either with `emacsql` or `emacsql-sqlite3`, two different authors), and it seems to be a Windows-only problem. 

Bottom line: I would not expect the problem to be addressed very quickly by either of the owners of the `emacsql` packages. More importantly, it's very difficult for the Org-roam project to address. As far as I know, @jethrokuan and @zaeph both use Linux, without access to Windows -- so they cannot see the problem themselves.



Hope this helps.

**Comment by notuntoward** [comment]
@nobiot, thanks for your detailed and helpful advice.  So far, setting ```
org-roam-db-update-method```  to 'immediate' makes this problem almost unnoticeable.

Do you know if the authors of `emacsql-sqlite3` are aware of this?

**Comment by nobiot** [comment]
Good to know it works for others. No, I don’t know if they are aware of the issue.

**Comment by gopar** [comment]
I'm having the same problems, except it always fails with the deleted buffer message.
Not sure what could be wrong. Was really excited in trying this out :/

**Comment by nobiot** [comment]
@gopar "Always fails with the deleted buffer message" is most likely a different issue. It might be better to open a new issue for tracking.

**Comment by gopar** [comment]
I figured out the issue. Turns out  I had `org-roam-db-location` set to a directory 🤦🏾 
Can't believe I never realized, until a few days later. Hopefully this might help someone not make a silly mistake like me.
